### PR TITLE
Decompress domain names in ResourceRecord data

### LINKED ISSUE #6936: DNS message parsing issue
#### Problem Description
In the current DNS implementation, we only decompress compression pointers from known formats in a DNS message such as domain names([example](https://github.com/mitmproxy/mitmproxy/blob/95bf441bc3a1dfcac2cb313cac8e72d3716a4bed/mitmproxy/net/dns/domain_names.py#L36)) due to which we corrupt compression pointers of data formats that we don't parse(for eg: `MX` records). This was supposed to be handled here:https://github.com/mitmproxy/mitmproxy/blob/95bf441bc3a1dfcac2cb313cac8e72d3716a4bed/mitmproxy/dns.py#L359 but isn't handled properly.

This issue can be fixed by decompressing all compression pointers properly

Note: Compression pointers (pair of bytes) are identified by leading `11` bits in the leading byte

#### Steps to reproduce the behavior:
1. Run `mitmproxy` in reverse DNS mode: `mitmproxy --mode reverse:dns://8.8.8.8@53`
2. Query for a record that returns data containing compression pointers: `dig zonetransfer.me MX @127.0.0.1`
3. `dig` returns a `bad label type` error

#### System Information
```
Mitmproxy: 11.0.0.dev
Python:    3.11.8
OpenSSL:   OpenSSL 3.2.1 30 Jan 2024
Platform:  Linux-6.8.11-amd64-x86_64-with-glibc2.38
```

### PR DESCRIPTION
**Pull request by errorxyz**
#### Description

Fixes #6936 

#### Checklist

 - [x] I have updated tests where applicable.
 - [x] I have added an entry to the CHANGELOG.

### CODE REVIEW

File: test/mitmproxy/test_dns.py [hunk 1]
```diff
@@ -186,8 +186,23 @@ def assert_eq(m: dns.Message, b: bytes) -> None:
             + b"\x00\x00\x03\x84\x00H\x07ns-1707\tawsdns-21\x02co\x02uk\x00\x11awsdns-hostmaster\x06amazon\xc0\x19\x00"
             + b"\x00\x00\x01\x00\x00\x1c \x00\x00\x03\x84\x00\x12u\x00\x00\x01Q\x80\x00\x00)\x02\x00\x00\x00\x00\x00\x00\x00"
         )
-        assert invalid_rr_domain_name.answers[0].data == b"\x99live\xc0\x12"
-
+        assert (
+            invalid_rr_domain_name.answers[0].data == b"\x99live\x06github\x03com\x00"
+        )
```
Comment (errorxyz) [inline test/mitmproxy/test_dns.py:191]: I didn't understand the purpose of this test🤔 and I'm not sure if this change is correct either

Comment (mhils) [reply test/mitmproxy/test_dns.py:191]: @meitinger, thoughts? 😅

File: mitmproxy/dns.py [hunk 2]
```diff
@@ -355,21 +355,37 @@ def unpack_rrs(
...
-                            ) = domain_names.unpack_from_with_compression(
-                                buffer, offset, cached_names
-                            )
-                            if rr_name_len == len_data:
-                                data = domain_names.pack(rr_name)
-                        except struct.error:
-                            pass
-                    section.append(ResourceRecord(name, type, class_, ttl, data))
+
+                    data = bytearray(buffer[offset:end_data])
```
[the remark this answers is not in the document]
Comment (errorxyz) [reply mitmproxy/dns.py:359]: bytearrays are mutable, using raw bytes for manipulation would be very slow

[the remark this answers is not in the document]
Comment (lups2000) [reply mitmproxy/dns.py:359]: Ahh I see good to know :) thanks!

**Comment by errorxyz** [comment]
> The code makes sense to me even though I don't have a 100% idea of what's going on 😅

ok so whats happening here is that DNS uses compression pointers to reduce the size of the message transmitted. DNS messages contain multiple ResourceRecords and these ResourceRecords contain the domain name followed by the data(such as IP address, mail address, etc.)  along with some other stuff. These domain names are usually represented by compression pointers (2 bytes in size) that are recognized by a leading `11` in their binary format followed by a pointer pointing to the data in the message. Now our current DNS implementation decompresses the compressed "domain name" part of the ResourceRecord but does not decompress compressed data in the data part of ResourceRecord due to which the size of the message changes and the compression pointers now point to garbage data. 

The issue is resolved by decompressing compression pointers in the data part of ResourceRecord as well. This is done by iterating through each byte and checking if it has a leading `0b11`, if so we try to decompress it and update it in the `data` variable. The offsets are adjusted accordingly. Now one caveat to this method is that, sometimes we might incorrectly identify data that would not contain compression pointers like A, AAAA records to contain compression pointers since their data might have a byte with a leading `0b11`. In such a case, we hope that it raises a `struct.error` somewhere (eg: pointer out of index, pointer points to corrupt data) in the code that is silenced by the `except struct.error` block. Hope this explanation clears things a bit.

**Comment by mhils** [comment]
> ok so whats happening here [...]

This is a super nice explanation, let's move your entire post into a source code comment! :)

File: test/mitmproxy/test_dns.py (hunk 1 above)
Comment (meitinger) [reply test/mitmproxy/test_dns.py:191]: The purpose of this test is the following: Since the original code performed a crude heuristic to check if a packed domain name is contained in the RR, it would indeed find the `0b110000000` byte (yes, pointers above an offset of 255 bytes don't trigger the old heuristic :) and try to expand the RR as domain name. But then it would encounter a `0x99` size specifier, which is not allowed (`0b10...` and `0x01...` are reserved) and abort all together. Since the new code would just ignore the first segment and still expand the `\xc0\x12` into github.com, updating the test is fine. (If we want that kind of behavior.)

**Comment by meitinger** [comment]
Thanks for spotting the issue and this PR, @errorxyz. (When I was writing the DNS message code, I didn't expect mitmproxy to handle/relaying anything other than A, AAAA, CNAME and _maybe_ TXT^^)

While I can appreciate the way you want to fix the code defect, may I offer a different approach:
How about we store the original bytes for everything past the `flags` in the header (i.e. for the variable length part), and use those bytes in `Message.packed` in case nothing got changed? I feel this might be the more robust thing to do than digging/decompressing deeper into bytes, we don't really know much about.

**Comment by errorxyz** [comment]
> in case nothing got changed

I did not use this method because things might get a bit complex if any parameter was to change. Let's see what @mhils has to say

**Comment by meitinger** [comment]
> I did not use this method because things might get a bit complex if any parameter was to change. Let's see what @mhils has to say

Valid point, but if we keep this behavior, might I suggest doing a `RR` type check, with the following intention:
- if known domain name (`CNAME`, ...): always call `domain_names.unpack_from_with_compression`
- if known binary (`A`, ...) don't do anything
- otherwise: use your method

Otherwise offset 168 will become very interesting in `A` messages of intranet DNS servers ;)

**Comment by mhils** [comment]
After digging a bit into this: This are more tricky and subtle than expected. For example, right now I think we will fail if the compressed message is slightly below the UDP payload limit, but after parsing and re-packing (where we don't compress), we're above the UDP payload limit. But that's an issue completely independent of this PR.

For compression, I think we should follow https://datatracker.ietf.org/doc/html/rfc3597#section-4, i.e. only decompress for the well-known types. My proposal would be we fix this, and then see how long our current implementation takes us without further complaints. _If_ we find that we are running into reliability issues, we should give https://crates.io/crates/hickory-proto a try. I built a quick prototype earlier ([1](https://github.com/mitmproxy/mitmproxy/compare/main...mhils:mitmproxy:dns-proto-experiment?expand=1), [2](https://github.com/mitmproxy/mitmproxy_rs/compare/main...mhils:mitmproxy_rs:dns-proto-experiment?expand=1)) to replace all parsing with that, and it works really quite well. We don't need to write our own parsing if there is a robust memory-safe alternative. :)

**Comment by meitinger** [comment]
> TL;DR: Let's decompress well known types only if that's easy. If we run into more limitations, Rust it is probably.

I can come up with a PR for that, which will also fix the invalidated pointers issue. (Basically what I said before, by keeping the original data around.)
This, however, will only work if both questions and answers are kept as-is for unknown RR types. If that is not the case, I'd say let's go for hickory. (Personally, I also wouldn't mind implementing more RR types in Python that use compressed domain names. With type 65 we actually got lucky, because
https://github.com/mitmproxy/mitmproxy/blob/27f990203a6db31b5b95d703e9d7f4fb6380183b/mitmproxy/dns.py#L120
wouldn't work with compressed rdata. :)

**Comment by mhils** [comment]
> (Basically what I said before, by keeping the original data around.)

Maybe I'm misunderstanding this, but this sounds like a (partial) rewrite that takes quite a bit of effort. At which point we should do Rust right away. My proposal would be to do the most minimal effort here to fix the bug @errorxyz encountered, then call it a day and see when it comes up again.

> [HTTPS records] wouldn't work with compressed rdata. :)

See https://datatracker.ietf.org/doc/html/rfc3597#section-4 I linked - unless I'm misunderstanding things, you can only have compression for types defined in RFC 1035 - no guessing required.

**Comment by meitinger** [comment]
I wouldn't call it a rewrite: https://github.com/mitmproxy/mitmproxy/compare/main...meitinger:mitmproxy:dns-fix?expand=1 :)
But we do have to change some playbook tests...
@errorxyz Do you think this change will fix your issue? I've tried your MX command and it works now.

**Comment by errorxyz** [comment]
Sorry for the delayed response @meitinger , I think what you're trying to do [here](https://github.com/mitmproxy/mitmproxy/compare/main...meitinger:mitmproxy:dns-fix?expand=1#diff-aae5ae0afc72a57598bfda9aaf2a6dc9aca959680e07ebb80b318c4233773099R68-R75) is set `_orig_bytes` of a single `ResourceRecord`/`Question` to `None` if any of its attributes was to change (and recalculate the raw bytes later). But I think we should recalculate the raw bytes of the entire message if a single attribute of any ResourceRecord was to change as it would corrupt compression pointers. Does this make sense?

**Comment by errorxyz** [comment]
> > For example, right now I think we will fail if the compressed message is slightly below the UDP payload limit, but after parsing and re-packing (where we don't compress), we're above the UDP payload limit
> 
> We can fix this by setting the `TC` bit if the new size is over the UDP limit. https://datatracker.ietf.org/doc/html/rfc7766#section-4

Turns out it isn't as simple as that, we can also have EDNS being used (which we don't parse) that allows UDP messages to be beyond 512 bytes (default UDP limit).
https://datatracker.ietf.org/doc/html/rfc6891
We could be strict and prevent these EDNS-specific OPT RR from being passed and truncate UDP messages at 512 bytes but I think for this PR we can leave things as they stand

**Review by mhils — APPROVED** [review]
Looks good to me, but I'm not in the weeds here. Does this look ok to ship to you, @meitinger? :)

**Comment by meitinger** [comment]
> Looks good to me, but I'm not in the weeds here. Does this look ok to ship to you, @meitinger? :)

These changes _could_ in theory cause issues with known domain name RRs, but since we don't have allowed DNS chars above 0xC0, are not required to do length checking, and unknown formats are broken anyways, we might as well merge it.
I'm just curious, why we shouldn't return unknown RRs as-is, as discussed above. I agree with @errorxyz, that the way I preserved the data could be more refined (i.e. if something has changed for _known_ RRs, recalculate everything); otherwise that would be a safer approach.

But I don't want to complicate things, so it's really up to you. This code will definitely work in most cases. :)

**Comment by errorxyz** [comment]
> I'm just curious, why we shouldn't return unknown RRs as-is

If I'm understanding things correctly, consider this example:
Say response has 3 records: 1st record is a NS record(known) and data contains example.com(compressed)
2nd record is a NS record(known) and data contains example2.com(uncompressed), 3rd record is an MX record(unknown) and its data has a compression pointer pointing to the data in the second record(example2.com).
Now if we decompress the 1st record, the compression pointer in the 3rd record would point to garbage data

> These changes could in theory cause issues with known domain name RRs

Do you have any example in mind?

**Comment by meitinger** [comment]
>  Now if we decompress the 1st record, the compression pointer in the 3rd record would point to garbage data

You're correct but getting `NS` and `MX` in the resource records would - in my opinion - already constitute as a query we shouldn't handle (since either for `ANY` et al or mixed questions), so if a hook decides to change something, it would be its responsibility to fix it.
Without a solution like yours, where we dig into unknown data, we simply can't do better.

> Do you have any example in mind?

Actually, the test that you needed to change is one example, where strictly going by [section 4.1.4](https://www.ietf.org/rfc/rfc1035.html#section-4.1.4) we should not decompress. Also, if there are consecutive pointers present, we now simply decompress. Same might happen for invalid domain names containing e.g. `Ä` (and a following letter leading to a valid pointer). But that's why I said in theory, because _valid_ (single domain name) RRs should be fine :)
